lib/nostrum/struct/emoji.ex

```diff
@@ -3,30 +3,111 @@ defmodule Nostrum.Struct.Emoji do
   Struct representing a Discord emoji.
   """
 
+  alias Nostrum.Struct.Snowflake
+  alias Nostrum.Struct.User
+  alias Nostrum.Util
+
+  defstruct [
+    :id,
+    :name,
+    :user,
+    :require_colons,
+    :managed,
+    :animated,
+    :roles
+  ]
+
+  @typedoc ~S"""
+  Emoji string to be used with the Discord API.
+
+  Some API endpoints take an `emoji`. If it is a custom emoji, it must be 
+  structured as `"id:name"`. If it is an unicode emoji, it can be structured 
+  as any of the following: 
+
+    * `"name"`
+    * A base 16 unicode emoji string.
+    * A URI encoded string.
+
+  We suggest library users to use the `to_api_name/1` function to get a 
+  `Nostrum.Struct.Emoji`'s emoji api name.
+
+  ## Examples
+
+      # Custom Emojis
+      "nostrum:431890438091489"
+
+      # Unicode Emojis
+      "≡ƒæì"
+      "\xF0\x9F\x98\x81"
+      URI.encode("\u2b50")
+
+      # All Emojis
+      emoji = %Nostrum.Struct.Emoji{id: 43819043108, name: "foxbot"}
+      Nostrum.Struct.Emoji.to_api_name(emoji)
+      "foxbot:43819043108"
+
+  """
+  @type emoji_api_name :: String.t
+
   @typedoc "Id of the emoji"
-  @type id :: integer
+  @type id :: Snowflake.t | nil
 
   @typedoc "Name of the emoji"
   @type name :: String.t
 
-  @typedoc """
-  Custom emoji string to be used with API
+  @typedoc "Roles this emoji is whitelisted to"
+  @type roles :: [Snowflake.t] | nil
 
-  Some API endpoints take an `emoji`.
-  The emoji should be structured as `"id:name"`.
-  """
-  @type custom_emoji :: String.t
+  @typedoc "User that created this emoji"
+  @type user :: User.t | nil
+
+  @typedoc "Whether this emoji must be wrapped in colons "
+  @type require_colons :: boolean | nil
+
+  @typedoc "Whether this emoji is managed"
+  @type managed :: boolean | nil
+
+  @typedoc "Whether this emoji is animated"
+  @type animated :: boolean | nil
 
   @type t :: %__MODULE__{
     id: id,
-    name: name
+    name: name,
+    roles: roles,
+    user: user,
+    require_colons: require_colons,
+    managed: managed,
+    animated: animated
   }
 
-  @derive [Poison.Encoder]
-  defstruct [
-    :id,
-    :name
-  ]
+  @doc """
+  Formats an emoji struct into its respective markdown.
+
+  ## Examples
+
+      emoji = %Nostrum.Struct.Emoji{id: 43819043108, name: "foxbot"}
+      Nostrum.Struct.Emoji.to_markdown(emoji)
+      "<:foxbot:43819043108>"
+  """
+  @spec to_markdown(t) :: String.t
```

Comment:
Correct me if I'm wrong, but the purpose of this here is for sending emojis in messages, yeah? It might be a good idea to state this or include an example of doing so.

As it stands I think `to_markdown` can be a little vague.

Reply:
I'll add an example in this upcoming commit.

I think `to_markdown` is vague as well. Maybe `format` would be a better name? Let me know your thoughts. Also, I was actually thinking about having the `String.Chars` protocol handle the markdown conversion. This would allow the emojis to be used in strings in a more natural method: 

```Elixir
emoji = %Nostrum.Struct.Emoji{id: 4314314, name: "craigpls"}

Kernel.to_string(emoji)
"<:craigpls:4314314>"

"I get to use this emoji now #{emoji}"
"I get to use this emoji now <:craigpls:4314314>"
```

And then we can always leave a markdown conversion function in case the user wanted a more explicit function call:

```Elixir
emoji = %Nostrum.Struct.Emoji{id: 4314314, name: "craigpls"}

md = Nostrum.Struct.Emoji.format(emoji)
"<:craigpls:4314314>"

"Got some new text #{md}"
"Got some new text <:craigpls:4314314>"
```

Let me know your thoughts.

Reply:
I think `format` or `format_emoji` would be good choices.

I like the protocol idea, lets go with that! Just make sure its documented properly.

Reply:
This next commit will reflect these changes.